// @blackwell-systems/gcf/dist/cjs/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.KIND_EXPAND = exports.KIND_ABBREV = void 0;
/**
 * KindAbbrev maps full kind names to short GCF abbreviations.
 */
exports.KIND_ABBREV = {
    function: 'fn',
    type: 'type',
    method: 'method',
    interface: 'iface',
    var: 'var',
    const: 'const',
    resource: 'resource',
    table: 'table',
    class: 'class',
    selector: 'selector',
    field: 'field',
    route_handler: 'route',
    external: 'ext',
    file: 'file',
    package: 'pkg',
    service: 'svc',
};
/**
 * KindExpand is the reverse of KindAbbrev: maps abbreviations back to full kind names.
 */
exports.KIND_EXPAND = {
    fn: 'function',
    type: 'type',
    method: 'method',
    iface: 'interface',
    var: 'var',
    const: 'const',
    resource: 'resource',
    table: 'table',
    class: 'class',
    selector: 'selector',
    field: 'field',
    route: 'route_handler',
    ext: 'external',
    file: 'file',
    pkg: 'package',
    svc: 'service',
};
