// @blackwell-systems/gcf/dist/cjs/stream.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StreamEncoder = void 0;
const constants_js_1 = require("./constants.js");
/**
 * StreamEncoder writes GCF output incrementally as symbols and edges arrive.
 * Zero buffering: each symbol/edge is written immediately. A trailer summary
 * is emitted on close() with the final counts.
 *
 * @example
 * ```ts
 * const chunks: string[] = [];
 * const enc = new StreamEncoder({ write: (s) => chunks.push(s) }, 'context_for_task', { tokenBudget: 5000 });
 * enc.writeSymbol({ qualifiedName: 'pkg.Auth', kind: 'function', score: 0.95, provenance: 'lsp', distance: 0 });
 * enc.writeEdge({ source: 'pkg.Server', target: 'pkg.Auth', edgeType: 'calls' });
 * enc.close();
 * ```
 */
class StreamEncoder {
    w;
    symIndex = new Map();
    nextID = 0;
    currentGroup = '';
    groupCounts = new Map();
    edgeCount = 0;
    edgesStarted = false;
    constructor(w, tool, opts = {}) {
        this.w = w;
        this.writeHeader(tool, opts);
    }
    writeHeader(tool, opts) {
        const parts = [`GCF profile=graph tool=${tool}`];
        if (opts.tokenBudget)
            parts.push(`budget=${opts.tokenBudget}`);
        if (opts.tokensUsed)
            parts.push(`tokens=${opts.tokensUsed}`);
        if (opts.packRoot)
            parts.push(`pack_root=${opts.packRoot}`);
        if (opts.session)
            parts.push('session=true');
        this.w.write(parts.join(' ') + '\n');
    }
    /**
     * Emit a symbol line immediately. Group headers are emitted automatically
     * when the distance changes.
     */
    writeSymbol(s) {
        const groupNames = ['targets', 'related', 'extended'];
        const groupName = s.distance < groupNames.length
            ? groupNames[s.distance]
            : `distance_${s.distance}`;
        if (groupName !== this.currentGroup) {
            this.w.write(`## ${groupName}\n`);
            this.currentGroup = groupName;
        }
        const id = this.nextID++;
        this.symIndex.set(s.qualifiedName, id);
        const kind = constants_js_1.KIND_ABBREV[s.kind] || s.kind;
        this.w.write(`@${id} ${kind} ${s.qualifiedName} ${s.score.toFixed(2)} ${s.provenance}\n`);
        this.groupCounts.set(groupName, (this.groupCounts.get(groupName) || 0) + 1);
    }
    /**
     * Emit an edge line immediately. The edges section header is emitted
     * automatically on the first edge (with [?] deferred count).
     * Source and target must reference previously-written symbols.
     */
    writeEdge(e) {
        const srcIdx = this.symIndex.get(e.source);
        const tgtIdx = this.symIndex.get(e.target);
        if (srcIdx === undefined || tgtIdx === undefined)
            return;
        if (!this.edgesStarted) {
            this.w.write('## edges [?]\n');
            this.edgesStarted = true;
        }
        let line = `@${tgtIdx}<@${srcIdx} ${e.edgeType}`;
        if (e.status && e.status !== 'unchanged') {
            line += ` ${e.status}`;
        }
        this.w.write(line + '\n');
        this.edgeCount++;
    }
    /**
     * Emit a bare reference for a previously-transmitted symbol (session mode).
     */
    writeBareRef(qname, distance) {
        const groupNames = ['targets', 'related', 'extended'];
        const groupName = distance < groupNames.length
            ? groupNames[distance]
            : `distance_${distance}`;
        if (groupName !== this.currentGroup) {
            this.w.write(`## ${groupName}\n`);
            this.currentGroup = groupName;
        }
        const id = this.nextID++;
        this.symIndex.set(qname, id);
        this.w.write(`@${id}  # previously transmitted\n`);
        this.groupCounts.set(groupName, (this.groupCounts.get(groupName) || 0) + 1);
    }
    /**
     * Emit the ##! summary trailer with final counts. Must be called after all
     * symbols and edges have been written.
     */
    close() {
        const deferredCounts = [];
        const groupOrder = ['targets', 'related', 'extended'];
        for (const g of groupOrder) {
            const c = this.groupCounts.get(g);
            if (c && c > 0)
                deferredCounts.push(c);
        }
        for (const [g, c] of this.groupCounts) {
            if (!groupOrder.includes(g) && c > 0)
                deferredCounts.push(c);
        }
        if (this.edgeCount > 0) {
            deferredCounts.push(this.edgeCount);
        }
        this.w.write(`##! summary symbols=${this.nextID} edges=${this.edgeCount} counts=${deferredCounts.join(',')}\n`);
    }
    /** Number of symbols written so far. */
    get symbolCount() { return this.nextID; }
    /** Number of edges written so far. */
    get edgeCount_() { return this.edgeCount; }
}
exports.StreamEncoder = StreamEncoder;
